File: shared/src/main/java/de/pnp/manager/model/character/Inventory.java
package de.pnp.manager.model.character;

import de.pnp.manager.model.ItemList;
import de.pnp.manager.model.item.IItem;
import org.jetbrains.annotations.NotNull;

import java.util.Collection;

public class Inventory extends ItemList implements IInventory {

    public static final Inventory EMPTY_INVENTORY = new Inventory(0, 0);

    protected double maxSize;
    protected int maxStackSize;

    public Inventory(double maxSize, int maxStackSize) {
        super();
        this.maxSize = maxSize;
        this.maxStackSize = maxStackSize;
    }

    public Inventory(double maxSize, int maxStackSize, IItem... items) {
        super(items);
        this.maxSize = maxSize;
        this.maxStackSize = maxStackSize;
    }

    public Inventory(double maxSize, int maxStackSize, Collection<IItem> items) {
        super(items);
        this.maxSize = maxSize;
        this.maxStackSize = maxStackSize;
    }

    @Override
    public boolean add(IItem item) {
        if (getCurrentSize() + item.getAmount() <= getMaxSize()) {
            if (contains(item)) {
                return super.add(item);
            } else {
                if (getCurrentStackSize() < getMaxStackSize()) {
                    return super.add(item);
                }
            }

        }
        return false;
    }

    @Override
    public boolean addAll(@NotNull Collection<? extends IItem> collection) {
        if (collection.stream().mapToDouble(IItem::getAmount).sum() + getCurrentSize() <= getMaxSize()) {
            long notContainedItemStacks = collection.stream().filter(i -> !contains(i)).count();

            if (getCurrentStackSize() + notContainedItemStacks <= getMaxStackSize()) {
                return super.addAll(collection);
            }
        }

        return false;
    }

    @Override
    public double getMaxSize() {
        return maxSize;
    }

    @Override
    public int getMaxStackSize() {
        return maxStackSize;
    }

    public void setMaxSize(double maxSize) {
        this.maxSize = maxSize;
    }

    public void setMaxStackSize(int maxStackSize) {
        this.maxStackSize = maxStackSize;
    }
}

File: app/src/main/java/de/pnp/manager/app/model/Inventory.kt
package de.pnp.manager.app.model

import de.pnp.manager.model.IItemList
import de.pnp.manager.model.character.IInventory
import de.pnp.manager.model.item.IItem

class Inventory : IInventory {
    private var maxSize: Int = 0
    private var maxStackSize: Int = 0
    private var itemList : IItemList? = null

    override fun getMaxSize(): Int {
        return maxSize
    }

    fun setMaxSize(maxSize: Int) {
        this.maxSize = maxSize
    }

    override fun getMaxStackSize(): Int {
        return maxStackSize
    }

    fun setMaxStackSize(maxStackSize : Int) {
        this.maxStackSize = maxStackSize
    }

    override fun getItems(): IItemList {
        return itemList!!
    }

    fun setItems(itemList : IItemList?) {
        this.itemList = itemList
    }
}